src/components/Player/AbstractPlayer.ts

```diff
@@ -0,0 +1,153 @@
+}
+
+
+
+// Given a Player object, calculate what state the AbstractPlayer component should have.
+// Bonus points if you like Continuation monads.
+function update_state(continuation: (state: AbstractPlayerState) => void, player: Player | void): void {
+    if (!player) {
+        continuation({
+            resolved: false,
+            guest: true,
+            username: "...",
+            icon: "",
+            country: "un",
+            short_rank: "",
+            long_rank: "",
+            rating: "",
+            className: ""
+        });
+    }
+    else if (is_guest(player)) {
+        continuation({
+            resolved: true,
+            guest: true,
+            username: player_name(player),
+            icon: "",
+            country: "un",
+            short_rank: "",
+            long_rank: "",
+            rating: "",
+            className: ""
+        });
+    }
+    else if (is_registered(player)) {
+        let suffix = (player.is.provisional ? "?" : "") + (player.is.timeout ? "T" : "");
+        let rating = Math.round(player.rating);
+        continuation({
+            resolved: true,
+            guest: false,
+            username: player_name(player),
+            icon: player.icon,
+            country: player.country || "un",
+            short_rank: "[" + rank_short_string(player.rank) + suffix + "]",
+            long_rank: rank_long_string(player.rank),
+            rating: !isNaN(rating) ? "" + rating : "...",
+            className: player_attributes(player).join(" ")
+        });
+    }
+}
+
+
+
+// An abstract superclass for implementing players in the user interface. This provides
+// functionality common to the Player and PlayerDetails components.
+export abstract class AbstractPlayer<S extends AbstractPlayerProperties, T extends AbstractPlayerState> extends React.PureComponent<S, T> {
+    protected player_id: number;
+    private subscribe: player_cache.Subscription;
+
+    constructor(props: S) {
+        super(props);
+
+        let callback: (player: Player) => void;
+        if (props.disableCacheUpdate) {
+            callback = (player) => undefined;
+        }
+        else {
+            callback = (player) => {
+                setTimeout(() => {
+                    // We must do this asynchronously to avoid an error
+                    // from within React. It complains bitterly if we
+                    // attempt to update our state during another
+                    // component's constructor. This can happen if
+                    // the other component's constructor causes our
+                    // player to be republished after we've been mounted.
+                    update_state(this.setState.bind(this), player);
+                });
+            };
+        }
+        this.subscribe = new player_cache.Subscription(callback);
+        this.setup(props, (state) => { this.state = state; });
+    }
+
+    componentWillReceiveProps(new_props) {
```

Comment:
This is probably where the redraw slowdown  elated to Player's is coming from, this should only doing updates if something changed (React doesn't try to make assumptions for us). 

We also need to implement `componentShouldUpdate` either here or in `Player` to check if we need to redraw based on new / old props and state